### PR TITLE
Update ccpp-physics. Make RRTMGP thread safe

### PR DESCRIPTION
**Pull request by dustinswales**
This PR contains several changes to the initialization and setup of the RRTMGP radiation scheme to allow for use across multiple openMP threads.
*) Moved Interstitial RRTMGP DDTs (ty_rrtmgp_gas_optics, ty_cloud_optics) to static fields defined during initialization in module memory.
*) Add "$omp critical" statements around the calling type-bound procedures during initialization.
*) Move allocation of ty_gas_conc from Interstitial to GFS_typedefs. Initialize ty_gas_concs for all blocks during initialization.

This issue was brought to our attention by @yangfanglin and Qingfu Liu at EMC while testing RRTMGP in high-resolution cases, which require multiple threads.

The GP regression tests on hera using Intel were successful.
There are no changes to the baselines

Waiting on
https://github.com/NCAR/ccpp-physics/pull/568
https://github.com/NOAA-EMC/fv3atm/pull/247

### CODE REVIEW

File: tests/tests/fv3_ccpp_gfs_v16_RRTMGP_2thrd [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                   RESTART/phy_data.tile2.nc \
+                   RESTART/phy_data.tile3.nc \
+                   RESTART/phy_data.tile4.nc \
+                   RESTART/phy_data.tile5.nc \
+                   RESTART/phy_data.tile6.nc"
+
+
+export_fv3
+export THRD=2
+export TPN=20
```
Comment (junwang-noaa) [inline tests/tests/fv3_ccpp_gfs_v16_RRTMGP_2thrd]: TPN=20 is for hera, maybe you can consider making changes like:
export TASKS=$TASKS_thrd
export TPN=$TPN_thrd
export INPES=$INPES_thrd
export JNPES=$JNPES_thrd

File: tests/tests/fv3_ccpp_gfs_v16_RRTMGP_2thrd [hunk 2]
```diff
@@ -0,0 +1,84 @@
+###############################################################################
+#
+#  FV3 CCPP GFS v16 w/ RRTMGP test using two openMP threads
+#
+###############################################################################
+
+export TEST_DESCR="Compare FV3 CCPP GFS v16 w/ RRTMGP using two threads results with previous trunk version"
+
+export CNTL_DIR=fv3_gfs_v16_RRTMGP_2thrd
```
Comment (DusanJovic-NOAA) [inline tests/tests/fv3_ccpp_gfs_v16_RRTMGP_2thrd]: If this test has the the same baseline outputs as `fv3_gfs_v16_RRTMGP` please change `CNTL_DIR` to `fv3_gfs_v16_RRTMGP`.

Comment (dustinswales) [reply tests/tests/fv3_ccpp_gfs_v16_RRTMGP_2thrd]: Will do

**Review by climbfuji — APPROVED** [review]

**Review by DusanJovic-NOAA — APPROVED** [review]

**Review by junwang-noaa — APPROVED** [review]
We still need orion log file, the code can be committed with that.

**Comment by climbfuji** [comment]
> We still need orion log file, the code can be committed with that.

@junwang-noaa CI tests passed. I'll quickly hop on orion and run the tests.

**Review by climbfuji — APPROVED** [review]

**Review by junwang-noaa — APPROVED** [review]

**Comment by pj-gdit** [comment]
I know this conversation is from some time ago but I'm wondering if the code modifications noted here are in the development branch.

We are running UFS on WCOSS2 HPE/Cray systems to evaluate RRTMGP radiation scheme and find that it fails in mo_gas_concentrat routine in longwave  code if more than 1 OMP thread is used.   shortwave code (and other RRTMGP routines executed before that) are OK.

I see, in the initial comment in this conversation, mention of adding OMP CRITICAL directives for thread safety (among other changes), however I do not see any OMP CRITICAL directives in the RRTMGP radiation code in the latest development branch.

Comments welcome.   Thanks,
Pete Johnsen

**Comment by dustinswales** [comment]
@pj-gdit It's been awhile, but I recall that we found a solution that didn't require adding OMP directives to the RRTMGp scheme.
There was some changes to RRTMGp a few months back, in which I added "blocking" to the longwave scheme. This is exactly as it sounds, where the block-size for the LW RRTMGP radiation is controlled at runtime via namelist. Since this is the piece you are having issues with, my instinct is that something isn't thread-safe with the new feature.

May I ask that you open a brief issue in the [UFS physics repository](https://github.com/ufs-community/ccpp-physics/issues)?
(Support for RRTMGp in the ccpp is limited at the moment, but if you open an issue, I will eventually get to it.)

**Pull request merged**